[expo-updates][ios] add error recovery manager

esamelson opened the pull request:
# Why

ENG-1493

This PR adds an Error Recovery manager class to iOS expo-updates. This class is intended to be used as a single instance per app, called into whenever JS encounters a fatal error or exception, and will coordinate the app's response.

The general control flow of this class is shown in this diagram:

<img width="1164" alt="Screen Shot 2021-09-13 at 10 49 42 AM" src="https://user-images.githubusercontent.com/19958240/133132221-e4487e1b-b9df-42fd-a022-fe85abdc6729.png">

It keeps track of a list of fallback "tasks" to try in sequence, and will continue moving down the list if more errors are reported, only crashing as a last resort if nothing else can be done.

This is hooked up to the rest of expo-updates in https://github.com/expo/expo/pull/14398, in case the implementation is helpful context.

# How

- Add new method to mark an update as FAILED in the local SQLite db
- When an error is reported, call this and then start the pipeline of tasks:

1. If a remote update is in progress (due to the app just having launched), wait up to 5s for it to finish, then try to launch that update if it's new. I chose 5s because it didn't seem like too ridiculously long to wait, and also we can expect that under reasonably good internet conditions an update will be able to finish downloading in this time.
2. If that fails, or if there's no new remote update, try to launch an older cached update
3. If that fails, try to launch the embedded update as a last resort
4. If that fails, crash

# Test Plan

Added unit tests for this manager for various possible scenarios of broken/fixed updates. Verified tests pass, and also ran some e2e tests manually with https://github.com/expo/expo/pull/14398.

# Future work

Possible future work might be to add a task to start a remote load if there isn't already one in progress, and to make the fallback tasks configurable, in case developers want to skip one or more of the possible steps.

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

jkhales commented on packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m:
```diff
@@ -0,0 +1,220 @@
+    case EXUpdatesErrorRecoveryTaskWaitForRemoteUpdate:
+      [self _waitForRemoteLoaderToFinish];
+      break;
+    case EXUpdatesErrorRecoveryTaskLaunchNew:
```
```suggestion
    // EXUpdatesErrorRecoveryTaskLaunchNew is called only after a new update is downloaded and added to the cache so it equivalent to EXUpdatesErrorRecoveryTaskLaunchCached
    case EXUpdatesErrorRecoveryTaskLaunchNew:
```

esamelson replied:
nice, added

jkhales commented on packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m:
```diff
@@ -0,0 +1,220 @@
+  });
+}
+
+- (void)_waitForRemoteLoaderToFinish
```
It seems like this function is supposed to get called twice on a successful remote load (otherwise we would never add `EXUpdatesErrorRecoveryTaskLaunchNew` to the pipeline) but I can't seem to find the code where this happens. What am I missing?

esamelson replied:
the way I have it set up is that if we call this function and a remote load is in progress, we then wait for `notifyNewRemoteLoadStatus` to be called when it's finished (which takes care of adding the task to the pipeline).

jkhales reviewed (commented):
Still trying to understand this PR fully, left a few Q's 🙏

ide commented on packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m:
```diff
@@ -0,0 +1,220 @@
+  dispatch_assert_queue(_errorRecoveryQueue);
+  if (_delegate.remoteLoadStatus == EXUpdatesRemoteLoadStatusLoading) {
+    _isWaitingForRemoteUpdate = YES;
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, _remoteLoadTimeout * NSEC_PER_MSEC), _errorRecoveryQueue, ^{
```
I left a comment [here](https://github.com/expo/expo/pull/14398#pullrequestreview-753467235) explaining why I am uncomfortable with using timer instead of subscribing specific events to trigger different phases of error recovery. This way we can reduce uncertainty in scenarios where the app is already in an uncertain state due to an error.

ide commented on packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m:
```diff
@@ -0,0 +1,220 @@
+
+- (void)handleError:(NSError *)error fromLaunchedUpdate:(nullable EXUpdatesUpdate *)launchedUpdate
+{
+  if (launchedUpdate) {
+    [self _markUpdateFailed:launchedUpdate];
+  }
+  [self _startPipelineWithEncounteredError:error];
```
Since _markUpdateFailed dispatches to the DB queue and then potentially dispatches to the error recovery queue, it's likely that _startPipelineWithEncounteredError will dispatch to the error recovery queue first. This creates a scenario where _markUpdateFailed removes EXUpdatesErrorRecoveryTaskLaunchCached from the task queue too late.

Possible fix is to give _markUpdateFailed a completion block.

esamelson replied:
Based on our conversation last week I decided this isn't worth the added complexity. If this database call fails, either the entire database is corrupt (in which case we'll emergency launch the embedded update next time), or this one call fails, in which case the worst thing that happens is we'll try to launch it again next time. The main priority with this whole feature is to allow the app to fix itself by downloading _new_ updates whenever possible, so I don't think this edge case with rolling back is worth about here. (I could be convinced otherwise)

(also this function has moved to https://github.com/expo/expo/pull/14398 -- https://github.com/expo/expo/pull/14398/files#diff-34b4cfcbaf8386b6601d2b279f4c38b67bbd8c5e962e276677a6176b879ac024R366)

esamelson committed:
[expo-updates][ios] add EXUpdatesErrorRecovery and tests

esamelson commented:
Based on a conversation with @ide last week, I've pushed a significant change to this stack of PRs which changes the behavior in the following ways:

- The error recovery manager listens for the `RCTContentDidAppearNotification`; after it has fired, we will mark the update as "successfully launched" and won't try to fall back to an older, outdated update. (old implementation relied on a timeout rather than events.)
- We still intercept the RCTFatal error handlers for a period of time after `RCTContentDidAppearNotification` is fired, to ensure that the app has a reasonable amount of time to check for a new update in case there is an error very soon after the initial render. In this case, however, we will NOT mark the update as failed (meaning we will keep trying to launch it until there's a new update) and we won't fall back to any older updates.
- Added new fields to the database to keep track of these states -- whether the update has successfully launched, failed to launch (i.e. an error was encountered before the initial render), and whether or not the update is outdated and should no longer be launched. [Because boolean fields are just integers anyway in SQLite, I made the success/fail columns counts rather than boolean values, in case this information is useful later on.]
- Added a new selection policy method to determine which updates should be marked "is_outdated" and should no longer be run. (This is to ensure that we don't accidentally roll back in any weird edge cases.)
- If there isn't already a remote update check in progress (or a new update downloaded) when an error is encountered, the error recovery manager will initiate one, unless updates are disabled. This is in line with the expected behavior in the managed/turtle-v1 workflow right now and the app.json key ON_ERROR_RECOVERY.
- Made some small changes/refactoring to allow for easier unit testing, and to allow us to fire any other error handlers that someone else might have already set on RCTFatal.

esamelson commented on packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m:
```diff
@@ -0,0 +1,220 @@
+- (instancetype)init
+{
+  return [self initWithErrorRecoveryQueue:dispatch_queue_create("expo.controller.errorRecoveryQueue", DISPATCH_QUEUE_SERIAL)
+                        remoteLoadTimeout:EXUpdatesErrorRecoveryRemoteLoadTimeoutMs];
```
We want this timeout to have a min value that would allow users to reasonably download an update within the time frame, since the app might crash after this timer runs out. I could see an argument for using something like `max(EXUpdatesLaunchWaitMs, EXUpdatesErrorRecoveryRemoteLoadTimeoutMs)` but I think that might be conflating two separate concepts and if we want this to be configurable we should just add a separate config value for it. I'm open to that idea but also think we could wait and see if it's wanted.

esamelson commented on packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m:
```diff
@@ -0,0 +1,220 @@
+    reason = ((NSException *)initialError).reason;
+    userInfo = ((NSException *)initialError).userInfo.mutableCopy;
+  } else {
+    NSAssert(NO, @"Shouldn't add object types other than NSError or NSException to _encounteredErrors");
```
yeah, this shouldn't ever happen unless it's our fault so was just intending this message for people developing expo-updates

esamelson commented on packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m:
```diff
@@ -0,0 +1,220 @@
+        [self->_pipeline removeObject:@(EXUpdatesErrorRecoveryTaskLaunchCached)];
+        [self _runNextTask];
+      } else {
+        self->_isRunning = NO;
```
> other than potentially a second pipeline from starting.

that is exactly what we want to block 😄 if RCTFatal somehow fires twice in quick succession we don't want to start two simultaneous update downloads, for example.

esamelson commented on packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m:
```diff
@@ -0,0 +1,220 @@
+  if (!success) {
+    [self _runNextTask];
+  } else {
+    self->_isRunning = NO;
```
I agree; that's partly what https://github.com/expo/expo/pull/14399 is for but I also think a full system for that is beyond the scope of this feature (at least for now).

esamelson commented on packages/expo-updates/ios/Tests/EXUpdatesErrorRecoveryTests.m:
```diff
@@ -0,0 +1,228 @@
+
+  NSError *mockError = mock([NSError class]);
+  [_errorRecovery handleError:mockError fromLaunchedUpdate:mock([EXUpdatesUpdate class])];
+  dispatch_sync(_testQueue, ^{}); // flush queue
```
Essentially, but we don't have a reference to all the blocks that are running on this queue, so in order to wait for them all to run, we take advantage of the fact that it's a serial queue and use `dispatch_sync`, which blocks the current queue until the empty block we're adding finishes executing.

esamelson committed:
PR comments

esamelson commented on packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m:
```diff
@@ -0,0 +1,220 @@
+  [self _startPipelineWithEncounteredError:exception];
+}
+
+- (void)notifyNewRemoteLoadStatus:(EXUpdatesRemoteLoadStatus)newStatus
```
yeah ([here's](https://github.com/expo/expo/pull/14398/files#diff-34b4cfcbaf8386b6601d2b279f4c38b67bbd8c5e962e276677a6176b879ac024R268) where it's used); I think of a callback slightly differently from a conceptual point of view (one class passes a method or block into another), which doesn't match this exactly. That's why I went with the naming I did, but I could be convinced to change it 😅

ide commented on packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m:
```diff
@@ -0,0 +1,285 @@
+      return obj.integerValue == EXUpdatesErrorRecoveryTaskWaitForRemoteUpdate || obj.integerValue == EXUpdatesErrorRecoveryTaskCrash;
+    }]].mutableCopy;
+  });
+  // wait 10s before unsetting error handlers; even though we won't try to
+  // relaunch if our handlers are triggered after now, we still want to give
+  // the EXUpdatesErrorRecoveryTaskWaitForRemoteUpdate task a reasonable
+  // amount of time to run and fetch a new update, in case there is one
```
If I understand correctly, the intent of this code is to wait after the app launches successfully by our definition (initial screen was rendered) so if there's an error in the next 10 seconds, we'll begin checking for + loading an update. Loading the update is given `_remoteLoadTimeout` milliseconds to run.

```suggestion
  // wait 10s before unsetting error handlers; even though we won't try to
  // relaunch if our handlers are triggered after now, we still want to give
  // the EXUpdatesErrorRecoveryTaskWaitForRemoteUpdate task a reasonable
  // window of time to start and check for a new update is there is one
```

ide commented on packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m:
```diff
@@ -0,0 +1,285 @@
+  } else {
+    _isWaitingForRemoteUpdate = YES;
+    [_delegate loadRemoteUpdate];
+    [self _runNextTask];
```
While the delegate is loading the update, won't this run the EXUpdatesErrorRecoveryTaskLaunchCached task right away? Two things: this case looks like it isn't waiting for the remote loader to finish (in a method named `_waitForRemoteLoaderToFinish`), and shouldn't this look for `checkAutomatically: ON_ERROR_RECOVERY` or similar instead of always looking for an update?

esamelson replied:
Re: running the next task, nice catch, that indeed shouldn't be here.

Re: `checkAutomatically: ON_ERROR_RECOVERY` -- this code won't run at all if updates are disabled entirely, but in the managed workflow we only allow two settings for `checkAutomatically`, `ALWAYS` and `ON_ERROR_RECOVERY`, and in both cases we would check for an update here. Do you think there's value in adding a third setting where updates are enabled but no automatic check happens, even in an error recovery situation?

esamelson committed:
[1] remove extra _runNextTask call

esamelson committed:
[1] fix tests

esamelson committed:
[1] PR feedback on comments

esamelson committed:
[1] fixes from testing (remove tasks instead of adding)

esamelson committed:
[skip ci][1] changelog
